### ISSUE TITLE
running error on local machibe

### ISSUE THREAD

**Issue by smazrouee** [niemasd/FAVITES#5]
running error on local machine

## FAVITES Version

latest (Feb 13th, 2018 )

## Environment

Mac OS

## Unexpected Behavior

{
    "ContactNetwork":                                   "NetworkX",
    "ContactNetworkGenerator":                          "RandonPartitionGraph",
    "Driver":                                           "Default",
    "EndCriteria":                                      "GEMF",
    "Logging":                                          "File",
    "NodeAvailability":                                 "Perfect",
    "NodeEvolution":                                    "None",
    "NumBranchSample":                                  "Single",
    "NumTimeSample":                                    "Once",
    "SeedSelection":                                    "Random",
    "SeedSequence":                                     "Random",
    "SequenceEvolution":                                "NoMutation",
    "seed_sequence_length":                             1,
    "Sequencing":                                       "Perfect",
    "SourceSample":                                     "Random",
    "TimeSample":                                       "End",
    "TransmissionNodeSample":                           "GEMF",
    "TransmissionTimeSample":                           "HIVARTGranichGEMF",
    "TreeNode":                                         "Simple",
    "TreeUnit":                                         "LogNormal",
    "end_time":                                         70,
    "rpg_sizes":                                        [20,20,20],
    "rpg_p_in":                                         0.1,
    "rpg_p_out":                                        0.002,
    "d_or_u":                                           "u",
    "gemf_path":                                        "GEMF",
    "grinder_path":                                     "grinder",
    "hiv_a1_to_a2":                                     4.333333333333333,
    "hiv_a1_to_d":                                      0,
    "hiv_a1_to_i1":                                     0.48,
    "hiv_a2_to_a3":                                     0,
    "hiv_a2_to_d":                                      0,
    "hiv_a2_to_i2":                                     0.48,
    "hiv_a3_to_a4":                                     0,
    "hiv_a3_to_d":                                      0,
    "hiv_a3_to_i3":                                     0,
    "hiv_a4_to_d":                                      0,
    "hiv_a4_to_i4":                                     0,
    "hiv_i1_to_a1":                                     0.5,
    "hiv_i1_to_d":                                      0,
    "hiv_i1_to_i2":                                     8.666666666666666,
    "hiv_i2_to_a2":                                     0.5,
    "hiv_i2_to_d":                                      0,
    "hiv_i2_to_i3":                                     0,
    "hiv_i3_to_a3":                                     0,
    "hiv_i3_to_d":                                      0,
    "hiv_i3_to_i4":                                     0,
    "hiv_i4_to_a4":                                     0,
    "hiv_i4_to_d":                                      0,
    "hiv_ns_to_d":                                      0,
    "hiv_ns_to_s":                                      999999,
    "hiv_s_to_d":                                       0,
    "hiv_s_to_i1_by_a1":                                0.005625,
    "hiv_s_to_i1_by_a2":                                0,
    "hiv_s_to_i1_by_a3":                                0,
    "hiv_s_to_i1_by_a4":                                0,
    "hiv_s_to_i1_by_i1":                                0.25,
    "hiv_s_to_i1_by_i2":                                0.05,
    "hiv_s_to_i1_by_i3":                                0,
    "hiv_s_to_i1_by_i4":                                0,
    "hiv_s_to_i1_seed":                                 0,
    "hmmemit_path":                                     "hmmemit",
    "java_path":                                        "java",
    "num_seeds":                                        1,
    "nw_rename_path":                                   "nw_rename",
    "out_dir":                                          "FAVITES_output_test3",
    "seed_population":                                  20,
    "seqgen_a_to_c":                                    1.765707,
    "seqgen_a_to_g":                                    9.587649,
    "seqgen_a_to_t":                                    0.691915,
    "seqgen_c_to_g":                                    0.863348,
    "seqgen_c_to_t":                                    10.282617,
    "seqgen_freq_a":                                    0.392,
    "seqgen_freq_c":                                    0.165,
    "seqgen_freq_g":                                    0.212,
    "seqgen_freq_t":                                    0.232,
    "seqgen_g_to_t":                                    1.0,
    "seqgen_gamma_shape":                               0.405129,
    "seqgen_num_gamma_rate_categories":                 "",
    "seqgen_path":                                      "seq-gen",
    "tree_rate_mean":                                   -5.5,
    "tree_rate_stdev":                                  0.2,
    "viral_sequence_type":                              "HIV1-B-DNA-POL-LITTLE",
    "vts_growthRate":                                   2.851904,
    "vts_model":                                        "logistic",
    "vts_n0":                                   1,
    "vts_t50":                                  -2,
    "num_cn_nodes"                                      "100"
    "ws_k"                                              "3"
    "ws_prob"                                           "0.02"
    "PostValidation":                                   "Dummy"
}

run_favites_docker.py -c Config-small-rp.json -u -v 
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 1316, in do_open
    encode_chunked=req.has_header('Transfer-encoding'))
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 1230, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 1276, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 1225, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 1017, in _send_output
    self.send(msg)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 955, in send
    self.connect()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 1391, in connect
    server_hostname=server_hostname)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 410, in wrap_socket
    _session=session
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 821, in __init__
    self.do_handshake()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 1075, in do_handshake
    self._sslobj.do_handshake()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 696, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:864)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/run_favites_docker.py", line 11, in <module>
    DOCKER_LATEST_TAG = sorted([t for t in urlopen("https://hub.docker.com/r/%s/tags/"%DOCKER_IMAGE).read().decode('utf-8').split('"tags":')[1].split(':')[-1][1:-2].replace('"','').split(',') if '.' in t])[-1]
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 525, in open
    response = self._open(req, data)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 543, in _open
    '_open', req)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 503, in _call_chain
    result = func(*args)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 1359, in https_open
    context=self._context, check_hostname=self._check_hostname)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 1318, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:864)>

**Comment by niemasd** [comment]
The error is the last line:

ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:864)

Is OpenSSL on your machine up to date? This is an error with your machine's Python 3, not with FAVITES

**Comment by smazrouee** [comment]
what should be the right version? below I copy the version of my openssl.  Any recent changes probably that enforces a certain version (it was working locally yesterday): 

OpenSSL 1.0.2n  7 Dec 2017
built on: reproducible build, date unspecified
platform: darwin64-x86_64-cc
options:  bn(64,64) rc4(ptr,int) des(idx,cisc,16,int) idea(int) blowfish(idx) 

 I even rebuilt my python after updating the openssl.

**Comment by smirarab** [comment]
Let's not close before the issue if fixed.

**Comment by niemasd** [comment]
I performed a quick Google search, and the following Stack Overflow link has a suggested solution:

https://stackoverflow.com/questions/42098126/mac-osx-python-ssl-sslerror-ssl-certificate-verify-failed-certificate-verify

**Comment by smazrouee** [comment]
Has there been any changes in docker script that requires changes/more
settings like this? now it's giving me docker (mounts denied) error. I know
how to fix it, but I am wondering whether there has been any changes (in
the code) that needs re-setting the current system which was working fine
until last night.

btw, we probably need to have a working/stable version that the users can
access (as the latest release using the -u option) which is separate from
what developer is working (/testing). So every time developer makes
(probable) changes such as fixing bugs, adding new capabilities a new
version/subversion should be released (and replaces the latest version).
Thanks

**Comment by niemasd** [comment]
> Has there been any changes in docker script that requires changes/more
settings like this?

Not that I'm aware of

> now it's giving me docker (mounts denied) error

What's the full error? In general, for all errors, it is helpful if you include the command you called and the entire error message/traceback

> I am wondering whether there has been any changes (in the code) that needs re-setting the current system which was working fine until last night

What do you mean by resetting the current system? Do you mean restarting the Docker service on Calab? If so, I don't think so: I was able to run `run_favites_docker.py` successfully just now on Calab

> btw, we probably need to have a working/stable version that the users can access (as the latest release using the -u option) which is separate from what developer is working (/testing)

Yes, that was the edit I made in the version of `run_favites_docker.py`: instead of using the `latest` tag (i.e., the developer bleeding-edge Docker image) as the default for the `-u` flag (which was how it was before), I now scan the Docker repository (using the `urllib` module, which is what was giving you the SSL certificate error) to figure out what the most recent *versioned* (i.e., official release, e.g. 1.0.3) Docker image exists, and then pulling that one as the default. So in other words, yes: the most recent `run_favites_docker.py -u` pulls the latest stable version of FAVITES

**Comment by smazrouee** [comment]
this thread is about the local machine not Caleb. And the error is docker
response error, mounts denied (which was working fine earlier, and with no
changes on the local machine is showing local file sharing problem). It's
an easy fix, though it's not clear why is it happening.

Seems there is a confusion here between changes in docker image (latest
version) and favorites_docker script latest version. Maybe I should issue
another thread for this to avoid the confusion.

**Comment by niemasd** [comment]
> this thread is about the local machine not Caleb. And the error is docker response error, mounts denied (which was working fine earlier, and with no changes on the local machine is showing local file sharing problem).

Can you post the exact command you ran and the full error? I.e., copy-and-paste them, not just a description

>  It's an easy fix, though it's not clear why is it happening.

What do you mean by this? What's the fix?

> Seems there is a confusion here between changes in docker image (latest version) and favorites_docker script latest version. Maybe I should issue another thread for this to avoid the confusion.

I'm not sure I understand what you mean by this. The state of `run_favites_docker.py` is completely independent of the state of the Docker images. `run_favites_docker.py` is just a script that pulls the latest Docker image and runs it, abstracting the actual command line calls to the `docker` executable from the user

**Comment by smazrouee** [comment]
Well... I figured it out. Seems like "run_favites_docker.py" script was changed last night (probably for the -o or something else). Consequently, I got the following errors: 
a. ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:864)
b. docker: Error response from daemon: Mounts denied:  The path ... is not shared from OS X and is not known to Docker.

And when I run the previous version of run_favites_docker.py, it runs without any problems. So I probably shouldn't have used the latest version of the python script (or should've tested it first). The errors are (almost) resolved, but I thought to leave an update here on what actually caused my local version those errors.

**Comment by niemasd** [comment]
> Well... I figured it out. Seems like "run_favites_docker.py" script was changed last night (probably for the -o or something else). Consequently, I got the following errors:
a. ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:864)
b. docker: Error response from daemon: Mounts denied: The path ... is not shared from OS X and is not known to Docker.

As I mentioned above, the update I made to `run_favites_docker.py` was to parse the FAVITES Docker repository (using `urllib`) to determine the most recent versioned Docker image, and also as I mentioned above, the issue that is causing the error you observed (`SSL: CERTIFICATE_VERIFY_FAILED`) is an issue with SSL verification on your Mac, and I had recommended you try the solution at the following Stack Overflow post:

https://stackoverflow.com/questions/42098126/mac-osx-python-ssl-sslerror-ssl-certificate-verify-failed-certificate-verify
